sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1261,6 +1397,138 @@ no queue state is changed. This design is because the queues are already in
 the state the user desires, so if the function threw an exception in this case,
 the application would likely swallow it and then proceed.
 
+=== External events
+
+The default behavior is that the event associated with a graph node cannot be 
+used to synchronize workload outside the graph.
+To overcome this limitation, users can define and use graph exernal events.
+These events can be used as regular sycl events. Users can therefore use them
+to manage the synchronizations between different graphs running concurrently 
+or between the graph's kernels and regular operations submit to a sycl queue.
+If a graph containing one or more external events is run multiple times, 
+the external events are automatically reset when the graph is resubmitted to 
+the queue.
```

Comment:
After reading through this, I'm think about whether we want to overload SYCL events with "external events" at all, as resetting an event isn't a concept in SYCL events and i'm not sure if we'll run into issues in the DPC++ runtime code because it assumes this can't happen. It's also not very intuitive to a user that they now need to maintain the knowledge of if a `sycl::event` variable is an external event, a graph-limited event, or a normal event.

An example of an alternative would be to define our own synchronization primitive that can be reset, let's call it
a "semaphore" for the sake of argument. This is what OpenCL have done in [cl_khr_semaphore](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_semaphore), then if we defined an API for it like:
```cpp
handler.wait_semaphore()
handler.signal_semaphore()
handler.reset_semaphore() // might not need this, not in cl_khr_semaphore
```
It could be used in CGF functions of both graph nodes and queue submission and would provide a simpler API for achieving the same goals. Need to check if it does indeed solve the same usecases, be implementable across backends, and it could probably be an extension in itself that we are compatible with (which is more work).

Reply:
So currently SYCL events don't have a concept of being reset, but I'm not sure this is necessarily a concern as prior to SYCL graphs a DAG could only be traversed once, so it only makes sense for events to be set once. But if a DAG (in the form of a `command_graph`) can be traversed multiple times I think it's a reasonable extension of the SYCL execution model that events within that graph are reset each time the graph is executed.

I do agree that it's a bit odd from the perspective of non-graph SYCL code for an event to be reset, though I would be wary of creating a new type as this could create duplicate mechanisms for doing the same thing in non-graph code. Perhaps we need a new concept to represent this to the existing SYCL execution model, such as a multi-shot event which can be complete and reset, and a way to query for this.

I think it could also be useful to have an API which can be used to query whether an event is external or graph-restricted, so that code which is simply passed an `event` object can create an "external event" if necessary.

Reply:
That's for the comment Gordon, it's good to know there's not an immediate objection to adding reset semantics to an event.